x/sync/workheap.go

```diff
@@ -98,13 +106,22 @@ func (wh *workHeap) MergeInsert(item *workItem) {
 	wh.sortedItems.DescendLessOrEqual(
 		searchItem,
 		func(beforeItem *heapItem) bool {
-			if item.localRootID == beforeItem.workItem.localRootID && bytes.Equal(beforeItem.workItem.end, item.start) {
-				// [beforeItem.start, beforeItem.end] and [item.start, item.end] are
-				// merged into [beforeItem.start, item.end]
-				beforeItem.workItem.end = item.end
-				beforeItem.workItem.priority = math.Max(item.priority, beforeItem.workItem.priority)
-				heap.Fix(&wh.innerHeap, beforeItem.heapIndex)
-				mergedBefore = beforeItem
+			if item.localRootID == beforeItem.workItem.localRootID {
+				beforeIsNothing := beforeItem.workItem.end.IsNothing()
+				itemIsNothing := item.start.IsNothing()
+
+				bothNothing := beforeIsNothing && itemIsNothing
+				bothHaveValue := !beforeIsNothing && !itemIsNothing
+				valuesMatch := bothHaveValue && bytes.Equal(beforeItem.workItem.end.Value(), item.start.Value())
+
+				if bothNothing || valuesMatch {
```

Comment:
I think this would be a lot more clear if we added a helper function for `Maybe[[]byte] Equals`